python/rikai/spark/functions/vision.py

```diff
@@ -280,3 +281,22 @@ def spectrogram_image(
         format=image_format,
         **image_kwargs,
     )
+
+
+@udf(returnType=ArrayType(IntegerType()))
+def img_cluster(img_list, threshold=0.5, img_shape=(300, 300)):
+    from sklearn.cluster import AgglomerativeClustering
+    from skimage.metrics import structural_similarity as ssim
+
+    if len(img_list) == 1:
+        return [0]
+    clustering_model = AgglomerativeClustering(
+        n_clusters=None, distance_threshold=threshold, linkage="ward"
+    )
+    frames = [np.resize(img.to_numpy(), img_shape) for img in img_list]
+    clustering_model.fit(
+        np.array(
+            [[ssim(f, ff, multichannel=True) for ff in frames] for f in frames]
```

Comment:
assuming ssim is commutative and self-ssim is a constant, you can save 1/2 of the computations here no?

Reply:
will do